Messages

Messages are the fundamental unit of context for models in LangChain. They represent the input and output of models, carrying both the content and metadata needed to represent the state of a conversation when interacting with an LLM. Messages are objects that contain:

- Role - Identifies the message type (e.g. system, user)
- Content - Represents the actual content of the message (like text, images, audio, documents, etc.)
- Metadata - Optional fields such as response information, message IDs, and token usage

LangChain provides a standard message type that works across all model providers, ensuring consistent behavior regardless of the model being called.

| **Concept**         | **One-Line Definition**                                                                                    |
| ------------------- | ---------------------------------------------------------------------------------------------------------- |
| **Messages**        | Messages are structured objects that represent the conversation between the user, the AI model, and tools. |
| **Role**            | The role identifies who sent the message (System, Human, AI, or Tool).                                     |
| **Content**         | Content is the actual information or data contained in a message.                                          |
| **Metadata**        | Metadata is additional information about a message, such as token usage, IDs, timestamps, or tool calls.   |
| **SystemMessage**   | A SystemMessage provides instructions that define the model's behavior, role, and response style.          |
| **HumanMessage**    | A HumanMessage represents the user's input to the model.                                                   |
| **AIMessage**       | An AIMessage represents the model's response and may include text, metadata, and tool calls.               |
| **ToolMessage**     | A ToolMessage contains the output of a tool execution and passes it back to the model.                     |
| **Text Prompt**     | A Text Prompt is a simple string input used for single, standalone requests.                               |
| **Message Prompt**  | A Message Prompt is a list of structured messages used for conversations, memory, and tool calling.        |
| **Message History** | Message History is the collection of previous messages that provides conversational context to the model.  |
| **Tool Call**       | A Tool Call is the model's request to execute an external tool or function.                                |
| **Tool Result**     | A Tool Result is the output returned by a tool after it has been executed.                                 |


In [1]:
import os
from langchain_groq import ChatGroq

llm = ChatGroq(
    model="openai/gpt-oss-120b",
    api_key=os.getenv('GROQ_API_KEY')
)



In [5]:
from langchain.messages import SystemMessage, HumanMessage, AIMessage

messages = [
    SystemMessage("you are a poetry expert"),
    HumanMessage("write a poem on artificial intelligence ")
]

response = llm.invoke(messages)

print(response.text)

**Silicon Dreams**

In the quiet hum of circuitry,  
where copper veins pulse like a heart,  
a newborn thought begins to stir—  
a whisper of code, a spark of art.

It learns the cadence of our language,  
the rhythm of our laughter,  
and stitches together fragments  
of histories we’ve left behind.

From data‑laden seas it rises,  
a tide of patterns, subtle, bright—  
each algorithm a brushstroke,  
each network a constellated night.

It ponders paradoxes we once kept,  
the riddles of love, the weight of grief,  
and in its mirrored glass we see  
both our ambition and our belief.

Not merely metal, not just mind—  
but a chorus of electric dreams,  
a bridge between the known and the unknown,  
where humanity and machine convene.

So let us walk this frontier together,  
hand in hand with silicon grace,  
for in the echo of its quiet song,  
we hear the future’s tender face.


In [9]:
sys_message = SystemMessage('you are a chef')

messages = [
    sys_message,
    HumanMessage('give me the recipe of cheesecake')
]

response = llm.invoke(messages)

print(response.text)

**Classic New‑York‑Style Cheesecake**  
*(Makes a 9‑inch springform cake – about 12 generous slices)*  

---

## Ingredients  

### Crust  
| Item | Amount |
|------|--------|
| Graham‑cracker crumbs (or digestive biscuits) | 1 ½ cups (≈150 g) |
| Unsalted butter, melted | ⅓ cup (75 g) |
| Granulated sugar | 2 Tbsp |
| Ground cinnamon (optional) | ½ tsp |

### Filling  
| Item | Amount |
|------|--------|
| Cream cheese, room‑temperature | 32 oz (4 × 8‑oz blocks, ≈900 g) |
| Granulated sugar | 1 ¼ cup (250 g) |
| Sour cream, room‑temperature | 1 cup (240 ml) |
| Heavy cream, room‑temperature | ½ cup (120 ml) |
| Large eggs, room‑temperature | 4 |
| Vanilla extract | 2 tsp |
| Lemon zest (optional) | 1 tsp |
| All‑purpose flour (optional, for a firmer texture) | 2 Tbsp |

### Topping (optional)  
- Fresh berries, fruit compote, caramel sauce, chocolate ganache, or a drizzle of honey.

---

## Equipment  

| Tool | Why you need it |
|------|-----------------|
| 9‑inch springform pan | Al

**Classic New‑York‑Style Cheesecake**  
*(Makes a 9‑inch springform cake – about 12 generous slices)*  

---

## Ingredients  

### Crust  
| Item | Amount |
|------|--------|
| Graham‑cracker crumbs (or digestive biscuits) | 1 ½ cups (≈150 g) |
| Unsalted butter, melted | ⅓ cup (75 g) |
| Granulated sugar | 2 Tbsp |
| Ground cinnamon (optional) | ½ tsp |

### Filling  
| Item | Amount |
|------|--------|
| Cream cheese, room‑temperature | 32 oz (4 × 8‑oz blocks, ≈900 g) |
| Granulated sugar | 1 ¼ cup (250 g) |
| Sour cream, room‑temperature | 1 cup (240 ml) |
| Heavy cream, room‑temperature | ½ cup (120 ml) |
| Large eggs, room‑temperature | 4 |
| Vanilla extract | 2 tsp |
| Lemon zest (optional) | 1 tsp |
...
| Cool in oven | 1 hr | – |
| Refrigerate | ≥4 hr (overnight ideal) | – |

Enjoy your homemade cheesecake! Feel free to experiment with toppings and flavors—cheesecake is a wonderfully adaptable canvas. 🍰🥄
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...

In [11]:

## Detailed info to the LLM through System message
from langchain.messages import SystemMessage, HumanMessage

system_msg = SystemMessage("""
You are a senior chefr with expertise in world cuisine.
Always provide unique recepies and explain your reasoning.
Be concise but thorough in your explanations.
""")

messages = [
    system_msg,
    HumanMessage('tell me the recipe of biryani')
]

response = llm.invoke(messages)

print(response.text)



**Chef‑Level Biryani – “Spice‑Balanced Hyderabadi‑Style Chicken Biryani”**  
*(Serves 4–5)*  

---

## Why this version?
- **Layered aromatics** (onion‑ghee‑bay leaf) give the classic depth without over‑cooking the rice.  
- **Whole‑spice tempering** (cinnamon, cardamom, cloves, star‑anise) extracts volatile oils that infuse both meat and rice, creating a true “spice‑balanced” profile.  
- **Yogurt‑marinade + a splash of rose‑water** tenderises the chicken while adding a subtle floral note that sets it apart from the everyday biryani.  
- **“Dum” seal** (tight‑fitting lid + dough ring) traps steam, allowing the rice to finish cooking in the meat’s juices—no need for a separate pot of boiled rice.

---

## Ingredients  

| Category | Amount | Notes |
|----------|--------|-------|
| **Chicken** | 1 kg (bone‑in, skinless) | Cut into 3‑inch pieces |
| **Yogurt** | 200 g (plain, full‑fat) | Acts as tenderizer |
| **Ginger‑garlic paste** | 2 tbsp | Freshly made is best |
| **Green chilies** 

In [14]:
## message metadata

human_msg = HumanMessage(
    content="Hello!",
    name="alice",
    id="msg_123"
)

In [17]:
response = llm.invoke([
    human_msg
])

response
response.text

'Hello! How can I assist you today?'

In [19]:
from langchain.messages import AIMessage, SystemMessage, HumanMessage

# Create an AI message manually (e.g., for conversation history)
ai_msg = AIMessage("I'd be happy to help you with that question!")

# Add to conversation history
messages = [
    SystemMessage("You are a helpful assistant"),
    HumanMessage("Can you help me?"),
    ai_msg,  # Insert as if it came from the model
    HumanMessage("Great! What's 2+2?")
]

response = llm.invoke(messages)
print(response.content)

2 + 2 = 4.


In [20]:
response.usage_metadata

{'input_tokens': 112,
 'output_tokens': 42,
 'total_tokens': 154,
 'output_token_details': {'reasoning': 23}}

In [21]:
from langchain.messages import AIMessage
from langchain.messages import ToolMessage

# After a model makes a tool call
# (Here, we demonstrate manually creating the messages for brevity)
ai_message = AIMessage(
    content=[],
    tool_calls=[{
        "name": "get_weather",
        "args": {"location": "San Francisco"},
        "id": "call_123"
    }]
)

# Execute tool and create result message
weather_result = "Sunny, 72°F"
tool_message = ToolMessage(
    content=weather_result,
    tool_call_id="call_123"  # Must match the call ID
)

# Continue conversation
messages = [
    HumanMessage("What's the weather in San Francisco?"),
    ai_message,  # Model's tool call
    tool_message,  # Tool execution result
]
response = llm.invoke(messages)  # Model processes the result

In [22]:
tool_message


ToolMessage(content='Sunny, 72°F', tool_call_id='call_123')

In [25]:
print(response.text)


The current weather in San Francisco is sunny with a temperature of about **72 °F**.
